src/object/cli_coll.c

```diff
@@ -101,14 +319,31 @@ obj_coll_oper_args_collapse(struct coll_oper_args *coa, uint32_t *size)
 		}
 	}
 
-	if (unlikely(coa->coa_dct_nr == 0))
-		/* If all shards are NONEXIST, then need not to send RPC(s). */
-		rc = 1;
-	else if (coa->coa_dct_cap > coa->coa_dct_nr)
-		/* Reset the other dct slots to avoid double free during cleanup. */
+	/* Reset the other dct slots to avoid double free during cleanup. */
+	if (coa->coa_dct_cap > coa->coa_dct_nr && coa->coa_dct_nr > 0)
 		memset(&coa->coa_dcts[coa->coa_dct_nr], 0,
 		       sizeof(*dct) * (coa->coa_dct_cap - coa->coa_dct_nr));
 
+	return 0;
+}
+
+static int
+obj_coll_oper_args_collapse(struct coll_oper_args *coa, struct dc_object *obj, uint32_t *size)
+{
+	int	rc;
+
+	if (coa->coa_sparse)
+		rc = obj_coll_collapse_tree(coa, size);
+	else
+		rc = obj_coll_collapse_array(coa, size);
+
+	if (rc >= 0) {
+		obj->cob_rank_nr = coa->coa_dct_nr;
+		/* If all shards are NONEXIST, then need not to send RPC(s). */
+		if (unlikely(coa->coa_dct_nr == 0))
+			rc = 1;
```

Comment:
same, is it really useful for rc 1 for caller?

Reply:
The same as above.